src/lib/renderComponent.tsx

```diff
@@ -2,50 +2,81 @@ import * as cx from 'classnames'
 import * as React from 'react'
 import { FelaTheme } from 'react-fela'
 
+import callable from './callable'
+import felaRenderer from './felaRenderer'
 import getClasses from './getClasses'
 import getElementType from './getElementType'
 import getUnhandledProps from './getUnhandledProps'
-import callable from './callable'
+import toCompactArray from './toCompactArray'
+
+import {
+  ComponentStyleFunctionParam,
+  ComponentVariablesInput,
+  ComponentVariablesObject,
+  IComponentPartClasses,
+  IComponentPartStylesInput,
+  IProps,
+  IThemeInput,
+  IThemePrepared,
+} from '../../types/theme'
 
 export interface IRenderResultConfig<P> {
   ElementType: React.ReactType<P>
-  rest: { [key: string]: any }
-  classes: { [key: string]: string }
+  classes: IComponentPartClasses
+  rest: IProps
 }
 
 export type RenderComponentCallback<P> = (config: IRenderResultConfig<P>) => any
 
+type IRenderConfigProps = {
+  [key: string]: any
+  variables?: ComponentVariablesInput
+  styles?: IComponentPartStylesInput
+}
+
 export interface IRenderConfig {
   className?: string
   defaultProps?: { [key: string]: any }
   displayName?: string
   handledProps: string[]
-  props: { [key: string]: any }
-  rules?: { [key: string]: Function }
-  variables?: (siteVariables: object) => object
+  props: IRenderConfigProps
 }
 
 const renderComponent = <P extends {}>(
   config: IRenderConfig,
   render: RenderComponentCallback<P>,
 ): React.ReactNode => {
-  const { className, defaultProps, displayName, handledProps, props, rules, variables } = config
+  const { className, defaultProps, displayName, handledProps, props } = config
 
   return (
     <FelaTheme
-      render={theme => {
-        const { siteVariables = {}, componentVariables = {}, renderer } = theme
-
+      render={({
+        siteVariables = {},
+        componentVariables = {},
+        componentStyles = {},
+        rtl = false,
+        renderer = felaRenderer,
+      }: IThemeInput | IThemePrepared = {}) => {
         const ElementType = getElementType({ defaultProps }, props)
         const rest = getUnhandledProps({ handledProps }, props)
-        const variablesFromFile = callable(variables)(siteVariables)
-        const variablesFromTheme = callable(componentVariables[displayName])(siteVariables)
-        const variablesFromProp = callable(props.variables)(siteVariables)
 
-        const mergedVariables = () =>
-          Object.assign({}, variablesFromFile, variablesFromTheme, variablesFromProp)
+        // Resolve variables for this component, allow props.variables to override
+        const variables: ComponentVariablesObject = {
+          ...callable(componentVariables[displayName])(siteVariables),
+          ...callable(props.variables)(siteVariables),
+        }
+
+        // Resolve styles using resolved variables, merge results, allow props.styles to override
+        const stylesForComponent = toCompactArray(componentStyles[displayName], props.styles)
```

Comment:
Can you be confident about the interface of styles/variables a component receives? It seems like you do a whole lot of truthiness checks when merging themes and styles at the provider level, and that would imply, to me, that consumers could be relieved of the "what if?" burden.

Reply:
All variables and styles can be objects or functions.  They are also optional.  We have confidence that they will either be missing, objects, or functions.  There are typings for these.

Could you expound on "consumers could be relieved of the "what if?" burden."?